### PR TITLE
[SofaExporter] Clean & test the exporter 

### PR DESCRIPTION
**Pull request by damienmarchal**
As pointed by @thomas-lemaire in #355 & #13 there are several issue on ObjExporter. 
In this PR I suggest to fix & unify the exporter (if you want to contribute...feel free to commit). 

Currently it:
- factor & unify shared code between OBJExporter, STLExporter and MeshExporter by inheriting from 
  BaseSimulationExporter
- adds a set of tests for OBJExporter, STLExporter and MeshExporter, the test coverage is not bad
- change of the API so that the ::write() function returns a boolean indicate the success or the failure
- add new possibility on the provided filename (to automatically create the missing path element)...
- report error & messages when the files cannot be created. 
- deprecation of the 'hard coded' interaction behavior... (I can explain if needed).
- fix crashes when missing topology

______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by thomas-lemaire** [comment]
Thanks Damien, I had a glance to #13 and I think @JeremieA proposal is nice to reduce bloat in sofa:  keeping the current MeshObj and its visitor mechanism to export obj for rendering (video...), so no need for a BaseSimulationExporter since only the MeshObjExporter would have this behavior.

On the other hand the MeshExporter component does not support any of the 2 formats we usually need, obj and stl... grrrr...

**Comment by damienmarchal** [comment]
Hi Thomas,

Many thanks for the feedback. 
I'm a bit confused because as far as I can read, Jeremie suggested to use OBJExporter not MeshObj to export visual things and I didn't changed this behavior. And there is no MeshObjExporter in sofa.

Here is the current situation in Sofa: 
In sofa/helper/io:
MeshObj  (a reading class, not a saving one) 
MeshStl  (a reading class, not a saving one) 

SofaExporter:
- MeshExporter (a writing component that focus on tetrahedral mesh forme as: VTKXML, VTK, Gmesh, Netgen, TetGen) All the format are coded in it. 
- OBJExporter (a writing component with a visitor that dump visual scene)
- STLExporter 
- VTKExporter 

As far as I can say the thee later are duplicating a lot of boiler plate code. The three reason why I created this BaseSimulationExporter was: to factorize theis code. To enforce a consistent behavior for each of them (eg having one component that create the missing "directory" while the other does not is too bad from a UX perspective). Finally it also help in writing the python binding because only one binding is needed instead of three. This is also consistent with what exists in SofaLoader where both MeshOBJLoader & MeshVTKLoader inherit from MeshLoader. 

My current thinking and target is that we need to have more symetrical implementation between SofaLoader and SofaExporter. 
Eg: 
in SofaExporter: 
- SceneExporter (rename or inherit from BaseSimulationExporter) 
      - SceneOBJExporter (the existing OBJExporter inheriting from SceneExporter)
      - SceneSTLExporter (the existing STLEXporter inheriting from SceneExporter)
      - ...
- MeshExporter (base class)
    - MeshTetGenExporter  (result from splitting the MeshExporter)
    - MeshVTKExporter (result from splitting the MeshExporter)

I don't have a lot of time to this so I prefer to stay focus on the "SceneExporter" (so STL & OBJ ones) and left the MeshExporter untouched for the moment. The rename with Scene* vs Mesh* is to clearly show the component that have a visitor in them. 

If you think it is ok I will do the cleaning of STLExporter as well...

**Comment by thomas-lemaire** [comment]
I agree with your way of splitting, except that the existing STLExporter is not a "SceneExporter" since it does not have a visitor mechanism, but rather works like the MeshExporter "family"

The only "SceneExporter" I see is the obj one for offline rendering

**Comment by damienmarchal** [comment]
You are right...This the BlenderExporter is also not having the visitor mechanism...

**Comment by thomas-lemaire** [comment]
anyway, this PR could be merged as is, maybe one can benefit from the factorized code in SceneExporter later

**Comment by damienmarchal** [comment]
I'm still making some factoring here...and just discovered that MeshExporter have its own way to generate the files names with kind of "wildcards". 

EDIT: some of them have no sense for scene specific exporter.

**Comment by damienmarchal** [comment]
Is VTKExporter equivalent to MeshExporter ? Can we remove it ?

**Comment by damienmarchal** [comment]
Hello @thomas-lemaire, @JeremieA, @guparan, @dequidt, @matthieu-nesme 

I made some work here, as I'm out of time for this task, I would appreciate if some of you can have a look for mistake and regression (It is not supposed to have any).

I was a bit annoyed with the "%" stuff in MeshExporter. 
A cleaner and more general approach should be made ...on day.

**Comment by matthieu-nesme** [comment]
Hi Damien,
it is a good project (the same for the loaders too). But it not easy to review.

For what I remember STLexporter is exporting from a list of pos/tri while the other exporters are exporting every VisualModels from their subgraph, isn't it?
I think both functionalities are interesting, and the baseexporter class could implement both?

**Comment by damienmarchal** [comment]
Hi Mathieu,

Thank for the feedback. 

STLExport & MeshExporter are exporting single object while OBJExporter is exporting a whole scene. 
I didn't change that because it would cause too much refactoring and it seems there is an agreement on that. 

In this PR what I did was just to factorize the existing quasi-duplicated functionalities into an unique class. So now the three objects have a consistent behavior regarding the filename & path creation as well as the attributes behavior. 

I'm very sorry about the difficulty to review this kind of changes... I think the best approach is not to use the diff which are too complex and instead review the resulting file which are (I hope) good looking. 

I also have added tests (+30) to check that things goes fine and help in future refactoring to detect regression. As you said one year ago ...we need more tests :)

**Comment by thomas-lemaire** [comment]
On vendredi 8 septembre 2017 09:20:09 CEST Matthieu Nesme wrote:
> Hi Damien,
> it is a good project (the same for the loaders too). But it not easy to
> review.
> 
> For what I remember STLexporter is exporting from a list of pos/tri while
> the other exporters are exporting every VisualModels from their subgraph,
> isn't it? 

no, see previous comment, only the obj exporter has this behavior to export a 
full simulation step in one obj and render it offline.

I think the "normal" behavior (no visitor) would better fit our current needs

thomas

> I think both functionalities are interesting, and the
> baseexporter class could implement both?

File: SofaKernel/framework/sofa/simulation/BaseSimulationExporter.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License *
+* for more details.                                                           *
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
```
Comment (guparan) [inline SofaKernel/framework/sofa/simulation/BaseSimulationExporter.cpp:21]: License header is duplicated here.

File: SofaKernel/framework/sofa/simulation/BaseSimulationExporter.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License *
+* for more details.                                                           *
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
```
Comment (guparan) [inline SofaKernel/framework/sofa/simulation/BaseSimulationExporter.h:21]: License header is duplicated here too.

**Review by guparan — CHANGES_REQUESTED** [review]

File: SofaKernel/framework/sofa/simulation/BaseSimulationExporter.h (hunk 2 above)
Comment (damienmarchal) [reply SofaKernel/framework/sofa/simulation/BaseSimulationExporter.h:21]: Shame on me for the duplicated licence...thanks for spotting that.

File: SofaKernel/framework/sofa/simulation/BaseSimulationExporter.cpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+#include <sofa/simulation/AnimateEndEvent.h>
+#include <sofa/helper/system/FileSystem.h>
+
+namespace sofa
+{
+
+namespace simulation
+{
+
+namespace _basesimulationexporter_
```
Comment (guparan) [inline SofaKernel/framework/sofa/simulation/BaseSimulationExporter.cpp:37]: I see you often use underscores in namespaces, could you explain me why?

File: SofaKernel/framework/sofa/simulation/BaseSimulationExporter.h (hunk 2 above)
Comment (damienmarchal) [reply SofaKernel/framework/sofa/simulation/BaseSimulationExporter.h:21]: Done.

File: SofaKernel/framework/sofa/simulation/BaseSimulationExporter.cpp (hunk 3 above)
Comment (damienmarchal) [reply SofaKernel/framework/sofa/simulation/BaseSimulationExporter.cpp:37]: Of course.

I'm using  dual namespace to separate what object are public and private in my .h/.cpp as this allow to employ the using keywords in .h without polluating the global namespaces. 

Eg: 
```cpp
namespace sofa  {
   namespace components {
       namespace myprivatens {
            using whatIWantWithoutPolluatingTheNS;
            class BlahBlah1{};
            class BlahBlah2{};
            class PrivateBlahBlah{};
     }
     /// Then "exxport" the public object
     using myprivatens::BlahBlah1;
     using myprivatens::BlahBlah2;
  }
} 
```

To make clear that the internal namespace is "private" I'm using a naming convention. 

The convention I know about this kind of thing is the qt one which look like: 
```cpp
namespace myfile
{
     namespace Private { ... };
}
```

Or the python way, that use underscore to indicate something is private. Looking more like 
```
namespace __myfile__ { ... };
```

I use arbitrarily to use the python convention.

Comment (guparan) [reply SofaKernel/framework/sofa/simulation/BaseSimulationExporter.cpp:37]: Very clear explanation, thanks! 👍

**Review by guparan — APPROVED** [review]

**Comment by guparan** [comment]
Warning: a bunch of unit tests are failing after merge on [ubuntu_clang-3.8_options-debug](https://ci.inria.fr/sofa-ci/job/ubuntu_clang-3.8_options-debug/353/testReport/)

**Pull request merged**